### PR TITLE
Updating related contents in other tabs won't refresh Preview panel and Page Editor #243

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

File: src/main/resources/assets/js/app/wizard/ContentWizardPanel.ts [hunk 1]
```diff
@@ -835,8 +841,13 @@ export class ContentWizardPanel
 
         // 1. template of the nearest site was updated
         // 2. nearest site was updated (app may have been added)
+        const nearestSiteChanged = (isPageTemplateUpdated && isUpdatedItemUnderSite) || (isSiteUpdated && isItemUnderUpdatedSite);
+
+        // 3. outbound dependency content has changed
+        return this.isOutboundDependencyUpdated(content).then(outboundDependencyUpdated => {
+            return nearestSiteChanged || outboundDependencyUpdated;
```
Comment (alansemenov) [inline src/main/resources/assets/js/app/wizard/ContentWizardPanel.ts]: With this approach we always make a request to check for outbound references, even when ``nearestSiteChanged`` is ``true`` and we don't really need to make this request.

What we need is this:
```Typescript
  const nearestSiteChanged = (isPageTemplateUpdated && isUpdatedItemUnderSite) || (isSiteUpdated && isItemUnderUpdatedSite);

  if (nearestSiteChanged) {
    return wemQ(true);
  }

  // 3. outbound dependency content has changed
  return this.isOutboundDependencyUpdated(content).then(outboundDependencyUpdated => outboundDependencyUpdated);

**Review by alansemenov — CHANGES_REQUESTED** [review]

File: src/main/resources/assets/js/app/browse/ContentBrowsePanel.ts [hunk 2]
```diff
@@ -687,6 +688,17 @@ export class ContentBrowsePanel
 
                 return previewRefreshed;
             });
+        }).then(() => {
+            if (!previewRefreshed) {
```
Comment (alansemenov) [inline src/main/resources/assets/js/app/browse/ContentBrowsePanel.ts]: Ok, this entire method is super-messy (not just the latest changes, but the old code as well).

Please do this:

1. Create a new method in ``Content`` class called ``isReferencedBy`` which will return a boolean promise. The method will take an array of either content ids or content objects and make a request to ContentQueryRequest (same as what now ``ContentQueryRequestHelper. anyIsOutboundDependency`` does).

2. Replace calls to ``ContentQueryRequestHelper. anyIsOutboundDependency`` with now beautifully looking calls to the new method, in both ``ContentBrowsePanel`` and ``ContentWizardPanel``:

``return previewItem.getModel().getContent().isReferencedBy(updatedContents.map(...));``

3. After that we no longer need ``ContentQueryRequestHelper``, so remove it.

4. Rename ``updatePreviewPanel`` to ``checkIfPreviewUpdateRequired`` and have it return a boolean promise of whether we need to update the preview (basically, return ``previewRefreshed``)

5. In this method remove all the calls to ``forcePreviewRerender``. It should only check whether we need to update the preview.

6. In ``handleContentUpdated``:
```Typescript
return this.doHandleContentUpdate(data).then((changed) => {
  this.checkIfPreviewUpdateRequired(data).then(previewUpdateRequired => previewUpdateRequired ? this.forcePreviewRerender() : null);
});
...
```

**Review by alansemenov — CHANGES_REQUESTED** [review]

File: src/main/resources/assets/js/app/browse/ContentBrowsePanel.ts (hunk 2 above)
Comment (alansemenov) [reply src/main/resources/assets/js/app/browse/ContentBrowsePanel.ts]: 7... And if you'll be keeping the ``previewRefreshed`` const then rename it to ``previewRefreshRequired``

**Comment by vbradnitski** [comment]
refactored.

**Comment by alansemenov** [comment]
Good stuff, but you missed one requirement in the description. Page Editor should not be refreshed if the page has been changed. Now, if I made changes to the page and an outbound reference was updated then the page will be refreshed and I will lose my changes.

**Comment by alansemenov** [comment]
Nize!

**Review by alansemenov — APPROVED** [review]

**Pull request merged**